File: app/src/main/java/com/mx/gillustrated/activity/AddCardActivity.java
package com.mx.gillustrated.activity;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.lang.ref.WeakReference;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

import com.mx.gillustrated.R;
import com.mx.gillustrated.adapter.SpinnerCommonAdapter;
import com.mx.gillustrated.common.MConfig;
import com.mx.gillustrated.util.CommonUtil;
import com.mx.gillustrated.util.PinyinUtil;
import com.mx.gillustrated.vo.CardInfo;
import com.mx.gillustrated.vo.CardTypeInfo;
import com.mx.gillustrated.vo.MatrixInfo;

import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Matrix;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.os.Message;
import android.provider.MediaStore;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnCheckedChanged;
import butterknife.OnClick;

public class AddCardActivity extends BaseActivity {

	private Button btnSave;
	private Spinner spinnerAttr;
	private Spinner spinnerLevel;
	private Spinner spinnerType;
	private EditText etId;
	private EditText etNid;
	private EditText etName;
	private EditText etFrontName;
	private EditText etCost;
	private EditText etHP;
	private EditText etAttack;
	private EditText etDefense;
	private ImageView ivNumber;
	private ImageView ivAll;
	private File m_fileNumber;
	private File m_fileAll;
	private Bitmap m_BitMapNumber;
	private Bitmap m_BitMapAll;
	private int mGameType;
	private File mImagesFileDir;
	private Button btnDelNumber;
	private Button btnDelAll;

	@BindView(R.id.chkAdjustImg)
	CheckBox chkAdjustImg;

	@BindView(R.id.tvAdjustImgTop)
	EditText tvAdjustImgTop;

	@BindView(R.id.tvAdjustImgBottom)
	EditText tvAdjustImgBottom;

	@BindView(R.id.tvAdjustImgLeft)
	EditText tvAdjustImgLeft;

	@BindView(R.id.tvAdjustImgRight)
	EditText tvAdjustImgRight;

	@OnClick(R.id.btnSaveMatrix)
	void onSaveMatrixBtnClick(){
		int top = Integer.parseInt(this.tvAdjustImgTop.getText().toString());
		int bottom = Integer.parseInt(this.tvAdjustImgBottom.getText().toString());
		int left = Integer.parseInt(this.tvAdjustImgLeft.getText().toString());
		int right = Integer.parseInt(this.tvAdjustImgRight.getText().toString());
		mSP.edit().putString(SHARE_IMAGES_MATRIX_NUMBER + mGameType, top + "," + bottom + "," + left + "," + right).commit();
		if(chkAdjustImg.isChecked()){
			try {
				this.showPicture();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}


	}

	@OnCheckedChanged(R.id.chkAdjustImg)
	void onCheckAdjustImgTopChanged(boolean checked){
		mSP.edit().putBoolean(SHARE_IMAGES_MATRIX + mGameType, checked).commit();
		try {
			this.showPicture();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}




	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_add);

		ButterKnife.bind(this);

		mGameType = getIntent().getIntExtra("game", 0);
				
		btnSave = (Button) findViewById(R.id.btnSave);
		btnSave.setOnClickListener(btnSaveClickListener);

		btnDelNumber = (Button) findViewById(R.id.btnDelNumber);
		btnDelNumber.setOnClickListener(btnDelNumberClickListener);
		btnDelAll = (Button) findViewById(R.id.btnDelAll);
		btnDelAll.setOnClickListener(btnDelAllClickListener);

		spinnerAttr = (Spinner) findViewById(R.id.spinnerAttr);
		spinnerLevel = (Spinner) findViewById(R.id.spinnerLevel);
		CommonUtil.setSpinnerItemSelectedByValue(spinnerLevel, "5");
		etNid = (EditText) findViewById(R.id.etDetailNid);
		etName = (EditText) findViewById(R.id.etDetailName);
		etFrontName = (EditText) findViewById(R.id.etDetailFrontName);
		etCost = (EditText) findViewById(R.id.etDetailCost);
		etId = (EditText) findViewById(R.id.etDetailId);
		etHP = (EditText) findViewById(R.id.etDetailHP);
		etAttack = (EditText) findViewById(R.id.etDetailAttack);
		etDefense = (EditText) findViewById(R.id.etDetailDefense);
		spinnerType  = (Spinner) findViewById(R.id.spinnerType);
		spinnerType.setSelection(0);
		spinnerType.setOnItemSelectedListener(onTypeSelectlistener);
		
		ivNumber = (ImageView) findViewById(R.id.imgWithNumber);
		ivAll = (ImageView) findViewById(R.id.imgAll);
		
		List<CardTypeInfo> cardTypes = mOrmHelper.getCardTypeInfoDao().queryForEq("game_type", mGameType);
		SpinnerCommonAdapter<CardTypeInfo> adapterName =
				new SpinnerCommonAdapter( this, cardTypes);
		spinnerAttr.setAdapter(adapterName);

		setImagesMatrixConfig();

		try {
			showPicture();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}

	}

	private void setImagesMatrixConfig(){
		String[] numbers = mSP.getString(SHARE_IMAGES_MATRIX_NUMBER + mGameType, "0,0,0,0").split(",");
		tvAdjustImgTop.setText(numbers[0]);
		tvAdjustImgBottom.setText(numbers[1]);
		tvAdjustImgLeft.setText(numbers[2]);
		tvAdjustImgRight.setText(numbers[3]);
		chkAdjustImg.setChecked(mSP.getBoolean(SHARE_IMAGES_MATRIX + mGameType, false));
	}

	private Bitmap getMatrixBitmap(Bitmap input){
		int top = Integer.parseInt(this.tvAdjustImgTop.getText().toString());
		int bottom = Integer.parseInt(this.tvAdjustImgBottom.getText().toString());
		int left = Integer.parseInt(this.tvAdjustImgLeft.getText().toString());
		int right = Integer.parseInt(this.tvAdjustImgRight.getText().toString());

		MatrixInfo matrixinfo = new MatrixInfo();
		matrixinfo.setY( top );
		matrixinfo.setX( left );
		matrixinfo.setHeight( top + bottom );
		matrixinfo.setWidth( left + right );

		return CommonUtil.cutBitmap(input, matrixinfo, true);
	}
	
	
	private void showPicture() throws FileNotFoundException, IOException {
		if (android.os.Environment.MEDIA_MOUNTED.equals(android.os.Environment
				.getExternalStorageState())) {
			File fileDir = new File(Environment.getExternalStorageDirectory(),
					MConfig.SRC_PATH);
			if (!fileDir.exists())
				fileDir = new File(Environment.getExternalStorageDirectory(),
						MConfig.SRC_PATH_SAMSUNG);
			if (!fileDir.exists())
				return;

			File file = new File(fileDir.getPath());
			File[] fs = file.listFiles();
			if(fs == null)
				return;
			Arrays.sort(fs, new Comparator<File>() {
				public int compare(File f1, File f2) {
					long diff = f1.lastModified() - f2.lastModified();
					if (diff > 0)
						return -1;
					else if (diff == 0)
						return 0;
					else
						return 1;
				}

				public boolean equals(Object obj) {
					return true;
				}

			});

			btnDelAll.setVisibility(View.GONE);
			btnDelNumber.setVisibility(View.GONE);
			ivAll.setImageBitmap(null);
			ivNumber.setImageBitmap(null);
			boolean isOrientation = mSP.getBoolean(SHARE_IMAGE_ORIENTATION + mGameType, false);
			for (int i = 0; i < fs.length; i++) {
				if (i == 2)
					break;

				Bitmap bmp = MediaStore.Images.Media.getBitmap(
						this.getContentResolver(), Uri.fromFile(fs[i]));
				if(this.chkAdjustImg.isChecked())
					bmp = getMatrixBitmap(bmp);

				if (i == 0) {
					m_fileAll = fs[i];
					m_BitMapAll = bmp;
					ivAll.setImageBitmap(isOrientation ? CommonUtil.rotatePic(bmp, 90) : bmp );
					btnDelAll.setVisibility(View.VISIBLE);
				} else {
					m_fileNumber = fs[i];
					m_BitMapNumber = bmp;
					ivNumber.setImageBitmap(isOrientation ? CommonUtil.rotatePic(bmp, 90) : bmp );
					btnDelNumber.setVisibility(View.VISIBLE);
				}
			}

		}

	}
	
	AdapterView.OnItemSelectedListener onTypeSelectlistener = new AdapterView.OnItemSelectedListener() {

		@Override
		public void onItemSelected(AdapterView<?> arg0, View arg1, int index,
				long arg3) {

			String[] array = getResources().getStringArray(R.array.addType);
			if( array[index].equals("更新附加图") || array[index].equals("更新数值图") ||  array[index].equals("新增单张图"))
			{
				ivNumber.setImageDrawable(null);
				btnDelNumber.setVisibility(View.GONE);
			}
		}

		@Override
		public void onNothingSelected(AdapterView<?> arg0) {

		}

	};


	View.OnClickListener btnDelNumberClickListener = new View.OnClickListener() {
		@Override
		public void onClick(View v) {
			CommonUtil.deleteImages(getBaseContext(), m_fileNumber);
			try {
				showPicture();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	};

	View.OnClickListener btnDelAllClickListener = new View.OnClickListener() {
		@Override
		public void onClick(View v) {
			CommonUtil.deleteImages(getBaseContext(), m_fileAll);
			try {
				showPicture();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	};

	View.OnClickListener btnSaveClickListener = new View.OnClickListener() {
		@Override
		public void onClick(View v) {

			mProgressDialog.show();
			new Thread() {
				public void run() {
					
					CardInfo card = new CardInfo();
					if(!"".equals(etNid.getText().toString().trim()))
						card.setNid(Integer.parseInt(etNid.getText().toString()));

					CardTypeInfo cardTypeInfo = (CardTypeInfo) spinnerAttr.getSelectedItem();
					card.setAttrId(cardTypeInfo.getId());
					card.setGameId(mGameType);
					card.setLevel(spinnerLevel.getSelectedItem().toString());
					card.setName(etName.getText().toString().trim());
					card.setPinyinName(PinyinUtil.convert(card.getName()));
					card.setFrontName(etFrontName.getText().toString().trim());
					card.setProfile("Y");
					if (!etCost.getText().toString().trim().equals(""))
						card.setCost(Integer.parseInt(etCost.getText()
								.toString()));
					else
						card.setCost(0);
					if (!etHP.getText().toString().trim().equals(""))
						card.setMaxHP(Integer.parseInt(etHP.getText()
								.toString()));
					if (!etAttack.getText().toString().trim().equals(""))
						card.setMaxAttack(Integer.parseInt(etAttack.getText()
								.toString()));
					if (!etDefense.getText().toString().trim().equals(""))
						card.setMaxDefense(Integer.parseInt(etDefense.getText()
								.toString()));
					int type = spinnerType.getSelectedItemPosition();
					
					mImagesFileDir = new File(
							Environment.getExternalStorageDirectory(),
							MConfig.SD_PATH + "/" + mGameType);
					if(!mImagesFileDir.exists()){
						mImagesFileDir.mkdirs();
			        }
					
					if(type == 0 || type == 1){
						mOrmHelper.getCardInfoDao().create(card);
						long newId = card.getId();
						if (newId != 0) {
							if (m_BitMapAll != null) {
								if(type == 0 && m_BitMapNumber != null )
								{
									createImages((int)newId, m_BitMapNumber, 1);
								}				
								if(type == 0 || type == 1)
								{
									if(m_BitMapNumber != null)
										createImages((int)newId, m_BitMapAll, type == 0 ? 2 : 1);
									else
										createImages((int)newId, m_BitMapAll, 1);
								}
								
								CommonUtil.deleteImages(AddCardActivity.this,
										m_fileAll);
								if(type == 0)
									CommonUtil.deleteImages(AddCardActivity.this,
											m_fileNumber);
							}
							Message msg = Message.obtain();
							msg.what = 1;
							msg.arg1 = (int) newId;
							addHandler.sendMessage(msg);
						}
					}else{
						// 更新数值图  更新附加图
						int id = Integer.parseInt(etId.getText().toString().trim());					
						int nextnum = getNextImagesIndex(id);
						File imageFile = new File(mImagesFileDir.getPath(),
								CommonUtil.getImageFrontName( id, 1));
						File nextFile = new File(mImagesFileDir.getPath(),
								CommonUtil.getImageFrontName( id, nextnum));
						if(type == 2){
							imageFile.renameTo(nextFile);
							createImages(id, m_BitMapAll, 1);
							card.setId(id);
							card.setLevel(null);
							card.setFrontName(null);
							card.setName(null);
							if(card.getCost() > 0 || card.getMaxHP() > 0 || card.getMaxAttack() > 0 || card.getMaxDefense() > 0)
								mOrmHelper.getCardInfoDao().update(card);
						}
						else if(type == 3)
							try {
								CommonUtil.exportImgFromBitmap(m_BitMapAll, nextFile);
							} catch (IOException e) {
								e.printStackTrace();
							} 
						
						CommonUtil.deleteImages(AddCardActivity.this,
								m_fileAll);
						addHandler.sendEmptyMessage(2);
					}
				}
			}.start();
		}
	};
	
	private void createImages(int id, Bitmap bitmap, int num){
		File imageFile;
		FileOutputStream bos;
		imageFile = new File(mImagesFileDir.getPath(),
				CommonUtil.getImageFrontName(id, num));
		try {
			bos = new FileOutputStream(imageFile);
			bitmap.compress(Bitmap.CompressFormat.JPEG,
					30, bos);
			bos.flush();
			bos.close();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
		
	}
	
	private int getNextImagesIndex(int id){
		int checknum = 3;
		while(true){
			File check = new File(mImagesFileDir.getPath(),
					CommonUtil.getImageFrontName( id, checknum));
			if(!check.exists()) 
				break;
			else
				checknum++;
		}
		return checknum;
	}
	

	private static class AddHandler extends Handler{

		private final WeakReference<AddCardActivity> mActivity;

		AddHandler( AddCardActivity activity){
			mActivity = new WeakReference<>(activity);
		}

		@Override
		public void handleMessage(Message msg) {
			super.handleMessage(msg);
			if (msg.what == 1) {
				Toast.makeText(mActivity.get(), "id: " + msg.arg1, Toast.LENGTH_SHORT).show();
				mActivity.get().forwardBack();
			} else if (msg.what == 2) {
				mActivity.get().forwardBack();
			}
		}
	}

	Handler addHandler = new AddHandler(this);

	
	private void forwardBack()
	{
		Intent intent = new Intent(AddCardActivity.this,
				MainActivity.class);
		intent.putExtra("game", mGameType);
		intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
		startActivity(intent);
		AddCardActivity.this.finish();
		mProgressDialog.dismiss();
	}
	
}

File: app/src/main/java/com/mx/gillustrated/activity/AddCardActivity.kt
package com.mx.gillustrated.activity

import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.lang.ref.WeakReference
import java.util.Arrays
import java.util.Comparator
import com.mx.gillustrated.R
import com.mx.gillustrated.adapter.SpinnerCommonAdapter
import com.mx.gillustrated.common.MConfig
import com.mx.gillustrated.util.CommonUtil
import com.mx.gillustrated.util.PinyinUtil
import com.mx.gillustrated.vo.CardInfo
import com.mx.gillustrated.vo.CardTypeInfo
import com.mx.gillustrated.vo.MatrixInfo
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.*
import android.provider.MediaStore
import android.util.SparseArray
import android.view.View
import android.widget.AdapterView
import android.widget.Toast
import androidx.core.util.forEach
import androidx.core.util.isEmpty
import androidx.core.util.valueIterator
import com.mx.gillustrated.databinding.ActivityAddBinding

@Suppress("DEPRECATION")
class AddCardActivity : BaseActivity() {

    private var mFileNumber: File? = null
    private var mFileAll: File? = null
    private var mBitMapNumber: Bitmap? = null
    private var mBitMapAll: Bitmap? = null

    private var mFileExtras:SparseArray<File> = SparseArray()
    private var mBitmapExtras:SparseArray<Bitmap> = SparseArray()

    private var mGameType: Int = 0
    private var mImagesFileDir: File? = null
    lateinit var binding:ActivityAddBinding

    private var onTypeSelectListener: AdapterView.OnItemSelectedListener = object : AdapterView.OnItemSelectedListener {

        override fun onItemSelected(arg0: AdapterView<*>, arg1: View, index: Int,
                                    arg3: Long) {
            if(index == 5){
                binding.spinnerExtra.visibility = View.VISIBLE
                binding.spinnerExtra.setSelection(1)
            }else{
                binding.spinnerExtra.visibility = View.GONE
            }
            try {
                showPicture()
            } catch (e: IOException) {
                e.printStackTrace()
            }

        }

        override fun onNothingSelected(arg0: AdapterView<*>) {

        }

    }



    private var btnDelNumberClickListener: View.OnClickListener = View.OnClickListener {
        deleteImages(mFileNumber!!, object :OnCallback{
            override fun deleted() {
                try {
                    showPicture()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        })
    }

    private var btnDelAllClickListener: View.OnClickListener = View.OnClickListener {
        deleteImages(mFileAll!!, object :OnCallback{
            override fun deleted() {
                try {
                    showPicture()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        })
    }

    private var btnSaveClickListener: View.OnClickListener = View.OnClickListener {
        mProgressDialog.show()
        object : Thread() {
            override fun run() {

                val card = CardInfo()
                if ("" != binding.etDetailNid.text.toString().trim { it <= ' ' })
                    card.nid = Integer.parseInt(binding.etDetailNid.text.toString())

                val cardTypeInfo = binding.spinnerAttr.selectedItem as CardTypeInfo
                card.attrId = cardTypeInfo.id
                card.gameId = mGameType
                card.level = binding.spinnerLevel.selectedItem.toString()
                card.name = binding.etDetailName.text.toString().trim { it <= ' ' }
                card.pinyinName = PinyinUtil.convert(card.name!!)
                card.frontName = binding.etDetailFrontName.text.toString().trim { it <= ' ' }
                card.profile = "Y"
                if (binding.etDetailCost.text.toString().trim { it <= ' ' } != "")
                    card.cost = Integer.parseInt(binding.etDetailCost.text
                            .toString())
                else
                    card.cost = 0
                if (binding.etDetailHP.text.toString().trim { it <= ' ' } != "")
                    card.maxHP = binding.etDetailHP.text
                            .toString().trim { it <= ' ' }
                if (binding.etDetailAttack.text.toString().trim { it <= ' ' } != "")
                    card.maxAttack = binding.etDetailAttack.text
                            .toString().trim { it <= ' ' }
                if (binding.etDetailDefense.text.toString().trim { it <= ' ' } != "")
                    card.maxDefense = binding.etDetailDefense.text
                            .toString().trim { it <= ' ' }
                if (binding.etDetailExtra1.text.toString().trim { it <= ' ' } != "")
                    card.extraValue1 = binding.etDetailExtra1.text
                            .toString().trim { it <= ' ' }
                if (binding.etDetailExtra2.text.toString().trim { it <= ' ' } != "")
                    card.extraValue2 = binding.etDetailExtra2.text
                            .toString().trim { it <= ' ' }

                val type = binding.spinnerType.selectedItemPosition

                mImagesFileDir = File(
                        Environment.getExternalStorageDirectory(),
                        MConfig.SD_PATH + "/" + mGameType)
                if (!mImagesFileDir!!.exists()) {
                    mImagesFileDir!!.mkdirs()
                }

                if (type == 0 || type == 1 || type == 4 || type == 5) { // type=4 && m_BitMapAll=null
                    mOrmHelper.cardInfoDao.create(card)
                    val newId = card.id.toLong()
                    if (newId != 0L) {
                        if (mBitMapAll != null) {
                            if (type == 0 && mBitMapNumber != null) {
                                createImages(newId.toInt(), mBitMapNumber!!, 1)
                            }
                            if (type == 0 || type == 1) {
                                if (mBitMapNumber != null)
                                    createImages(newId.toInt(), mBitMapAll!!, if (type == 0) 2 else 1)
                                else
                                    createImages(newId.toInt(), mBitMapAll!!, 1)
                            }

                            if (type == 0)
                                deleteImages(listOf(mFileNumber!!, mFileAll!!), object :OnCallback{
                                    override fun deleted() {
                                        val msg = Message.obtain()
                                        msg.what = 1
                                        msg.arg1 = newId.toInt()
                                        addHandler.sendMessage(msg)
                                    }

                                })
                            else if(type == 1)
                                deleteImages(mFileAll!!, object :OnCallback{
                                    override fun deleted() {
                                        val msg = Message.obtain()
                                        msg.what = 1
                                        msg.arg1 = newId.toInt()
                                        addHandler.sendMessage(msg)
                                    }

                                })
                        }else if(!mBitmapExtras.isEmpty()) { //0 - 5
                            val total = mBitmapExtras.size()
                            mBitmapExtras.forEach { key, value ->
                                createImages(newId.toInt(), value, total - key)
                            }
                            val files = mutableListOf<File>()
                            mFileExtras.forEach { key, value ->  files.add(value)}
                            deleteImages(files, object :OnCallback{
                                override fun deleted() {
                                    val msg = Message.obtain()
                                    msg.what = 1
                                    msg.arg1 = newId.toInt()
                                    addHandler.sendMessage(msg)
                                }

                            })
                        }else {
                            val msg = Message.obtain()
                            msg.what = 1
                            msg.arg1 = newId.toInt()
                            addHandler.sendMessage(msg)
                        }
                    }
                } else if(type == 2 || type == 3) {
                    // 更新数值图  更新附加图
                    val id = Integer.parseInt(binding.etDetailId.text.toString().trim { it <= ' ' })
                    val nextnum = getNextImagesIndex(id)
                    val imageFile = File(mImagesFileDir!!.path,
                            CommonUtil.getImageFrontName(id, 1))
                    val nextFile = File(mImagesFileDir!!.path,
                            CommonUtil.getImageFrontName(id, nextnum))
                    if (type == 2) {
                        imageFile.renameTo(nextFile)
                        createImages(id, mBitMapAll!!, 1)
                        card.id = id
                        card.level = null
                        card.frontName = null
                        card.name = null
                        if (card.cost > 0 || card.maxHP != null || card.maxAttack != null
                                || card.maxDefense != null || card.extraValue1 != null || card.extraValue2 != null )
                            mOrmHelper.cardInfoDao.update(card)
                    } else if (type == 3)
                        try {
                            CommonUtil.exportImgFromBitmap(mBitMapAll!!, nextFile)
                        } catch (e: IOException) {
                            e.printStackTrace()
                        }

                    deleteImages(mFileAll!!, object :OnCallback{
                        override fun deleted() {
                            addHandler.sendEmptyMessage(2)
                        }
                    })

                }
            }
        }.start()
    }

    internal var addHandler: Handler = AddHandler(this)


    companion object {

        private class AddHandler constructor(activity: AddCardActivity) : Handler(Looper.getMainLooper()) {

            private val mActivity: WeakReference<AddCardActivity> = WeakReference(activity)

            override fun handleMessage(msg: Message) {
                super.handleMessage(msg)
                if (msg.what == 1) {
                    Toast.makeText(mActivity.get(), "id: " + msg.arg1, Toast.LENGTH_SHORT).show()
                    mActivity.get()?.forwardBack()
                } else if (msg.what == 2) {
                    mActivity.get()?.forwardBack()
                }
            }
        }

    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAddBinding.inflate(layoutInflater)
        setContentView(binding.root)

        mGameType = intent.getIntExtra("game", 0)


        binding.btnSave.setOnClickListener(btnSaveClickListener)
        binding.btnDelNumber.setOnClickListener(btnDelNumberClickListener)
        binding.btnDelAll.setOnClickListener(btnDelAllClickListener)
        CommonUtil.setSpinnerItemSelectedByValue(binding.spinnerLevel, "5")

        val cardTypes = mOrmHelper.cardTypeInfoDao.queryForEq("game_type", mGameType)
        val adapterName = SpinnerCommonAdapter(this, cardTypes)
        binding.spinnerAttr.adapter = adapterName
        setImagesMatrixConfig()
        binding.spinnerType.onItemSelectedListener = onTypeSelectListener
        binding.spinnerType.setSelection(4)

        binding.btnSaveMatrix.setOnClickListener {
            val top = Integer.parseInt(binding.tvAdjustImgTop.text.toString())
            val bottom = Integer.parseInt(binding.tvAdjustImgBottom.text.toString())
            val left = Integer.parseInt(binding.tvAdjustImgLeft.text.toString())
            val right = Integer.parseInt(binding.tvAdjustImgRight.text.toString())
            mSP.edit().putString(SHARE_IMAGES_MATRIX_NUMBER + mGameType, "$top,$bottom,$left,$right").apply()
            if (binding.chkAdjustImg.isChecked) {
                try {
                    this.showPicture()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
        binding.chkAdjustImg.setOnCheckedChangeListener { _, isChecked ->
            mSP.edit().putBoolean(SHARE_IMAGES_MATRIX + mGameType, isChecked).apply()
            try {
                this.showPicture()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }

    }

    private fun setImagesMatrixConfig() {
        val numbers = mSP.getString(SHARE_IMAGES_MATRIX_NUMBER + mGameType, "0,0,0,0")!!.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        binding.tvAdjustImgTop.setText(numbers[0])
        binding.tvAdjustImgBottom.setText(numbers[1])
        binding.tvAdjustImgLeft.setText(numbers[2])
        binding.tvAdjustImgRight.setText(numbers[3])
        binding.chkAdjustImg.isChecked = mSP.getBoolean(SHARE_IMAGES_MATRIX + mGameType, false)
    }

    private fun getMatrixBitmap(input: Bitmap): Bitmap {
        val top = Integer.parseInt(binding.tvAdjustImgTop.text.toString())
        val bottom = Integer.parseInt(binding.tvAdjustImgBottom.text.toString())
        val left = Integer.parseInt(binding.tvAdjustImgLeft.text.toString())
        val right = Integer.parseInt(binding.tvAdjustImgRight.text.toString())

        val matrixInfo = MatrixInfo()
        matrixInfo.y = top
        matrixInfo.x = left
        matrixInfo.height = top + bottom
        matrixInfo.width = left + right

        return CommonUtil.cutBitmap(input, matrixInfo, true)
    }


    @Throws(FileNotFoundException::class, IOException::class)
    private fun showPicture() {
        binding.btnDelAll.visibility = View.GONE
        binding.btnDelNumber.visibility = View.GONE
        binding.btnDelExtra1.visibility = View.GONE
        binding.btnDelExtra2.visibility = View.GONE
        binding.btnDelExtra3.visibility = View.GONE
        binding.btnDelExtra4.visibility = View.GONE
        binding.btnDelExtra5.visibility = View.GONE
        binding.btnDelExtra6.visibility = View.GONE
        binding.imgAll.setImageBitmap(null)
        binding.imgWithNumber.setImageBitmap(null)
        binding.imgExtra1.setImageBitmap(null)
        binding.imgExtra2.setImageBitmap(null)
        binding.imgExtra3.setImageBitmap(null)
        binding.imgExtra4.setImageBitmap(null)
        binding.imgExtra5.setImageBitmap(null)
        binding.imgExtra6.setImageBitmap(null)
        mBitMapNumber = null
        mBitMapAll = null
        mFileAll = null
        mFileNumber = null
        mFileExtras = SparseArray()
        mBitmapExtras = SparseArray()
        binding.llImageExtra.visibility = View.GONE
        val type = binding.spinnerType.selectedItemPosition
        if (type == 4)
            return
        if (type == 5)
            binding.llImageExtra.visibility = View.VISIBLE
        if (Environment.MEDIA_MOUNTED == Environment
                        .getExternalStorageState()) {
            var fileDir = File(Environment.getExternalStorageDirectory(),
                    MConfig.SRC_PATH)
            if (!fileDir.exists())
                fileDir = File(Environment.getExternalStorageDirectory(),
                        MConfig.SRC_PATH_SAMSUNG)
            if (!fileDir.exists())
                return

            val file = File(fileDir.path)
            val fs = file.listFiles() ?: return
            Arrays.sort(fs, object : Comparator<File> {
                override fun compare(f1: File, f2: File): Int {
                    val diff = f1.lastModified() - f2.lastModified()
                    return when {
                        diff > 0 -> -1
                        diff == 0L -> 0
                        else -> 1
                    }
                }

                override fun equals(other: Any?): Boolean {
                    return true
                }

            })
            val isOrientation = mSP.getBoolean(SHARE_IMAGE_ORIENTATION + mGameType, false)
            val extraTotal =  binding.spinnerExtra.selectedItem.toString().toInt()
            for (i in fs.indices) {
                if (i == 1 && (type == 1 || type == 2 || type == 3))
                    break
                if (i == 2 && type == 0)
                    break
                if (i == extraTotal && type == 5)
                    break
                var bmp = MediaStore.Images.Media.getBitmap(
                        this.contentResolver, Uri.fromFile(fs[i]))
                if (binding.chkAdjustImg.isChecked)
                    bmp = getMatrixBitmap(bmp)
                if(type != 5){
                    if (i == 0) {
                        mFileAll = fs[i]
                        mBitMapAll = bmp
                        binding.imgAll.setImageBitmap(if (isOrientation) CommonUtil.rotatePic(bmp, 90) else bmp)
                        binding.btnDelAll.visibility = View.VISIBLE
                    } else if (i == 1) {
                        mFileNumber = fs[i]
                        mBitMapNumber = bmp
                        binding.imgWithNumber.setImageBitmap(if (isOrientation) CommonUtil.rotatePic(bmp, 90) else bmp)
                        binding.btnDelNumber.visibility = View.VISIBLE
                    }
                }else{
                    when(i){
                        0 -> {
                            mFileExtras.put(0, fs[i])
                            mBitmapExtras.put(0, bmp)
                            binding.imgExtra1.setImageBitmap(if (isOrientation) CommonUtil.rotatePic(bmp, 90) else bmp)
                            binding.btnDelExtra1.visibility = View.VISIBLE
                        }
                        1 -> {
                            mFileExtras.put(1, fs[i])
                            mBitmapExtras.put(1, bmp)
                            binding.imgExtra2.setImageBitmap(if (isOrientation) CommonUtil.rotatePic(bmp, 90) else bmp)
                            binding.btnDelExtra2.visibility = View.VISIBLE
                        }
                        2 -> {
                            mFileExtras.put(2, fs[i])
                            mBitmapExtras.put(2, bmp)
                            binding.imgExtra3.setImageBitmap(if (isOrientation) CommonUtil.rotatePic(bmp, 90) else bmp)
                            binding.btnDelExtra3.visibility = View.VISIBLE
                        }
                        3 -> {
                            mFileExtras.put(3, fs[i])
                            mBitmapExtras.put(3, bmp)
                            binding.imgExtra4.setImageBitmap(if (isOrientation) CommonUtil.rotatePic(bmp, 90) else bmp)
                            binding.btnDelExtra4.visibility = View.VISIBLE
                        }
                        4 -> {
                            mFileExtras.put(4, fs[i])
                            mBitmapExtras.put(4, bmp)
                            binding.imgExtra5.setImageBitmap(if (isOrientation) CommonUtil.rotatePic(bmp, 90) else bmp)
                            binding.btnDelExtra5.visibility = View.VISIBLE
                        }
                        5 -> {
                            mFileExtras.put(5, fs[i])
                            mBitmapExtras.put(5, bmp)
                            binding.imgExtra6.setImageBitmap(if (isOrientation) CommonUtil.rotatePic(bmp, 90) else bmp)
                            binding.btnDelExtra6.visibility = View.VISIBLE
                        }
                    }
                }
            }

        }

    }

    private fun createImages(id: Int, bitmap: Bitmap, num: Int) {
        val bos: FileOutputStream
        val imageFile = File(mImagesFileDir!!.path,
                CommonUtil.getImageFrontName(id, num))
        try {
            bos = FileOutputStream(imageFile)
            bitmap.compress(Bitmap.CompressFormat.JPEG,
                    30, bos)
            bos.flush()
            bos.close()
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    private fun getNextImagesIndex(id: Int): Int {
        var checkNum = 3
        while (true) {
            val check = File(mImagesFileDir!!.path,
                    CommonUtil.getImageFrontName(id, checkNum))
            if (!check.exists())
                break
            else
                checkNum++
        }
        return checkNum
    }

    private fun forwardBack() {
        val intent = Intent(this@AddCardActivity,
                MainActivity::class.java)
        intent.putExtra("game", mGameType)
        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
        startActivity(intent)
        this@AddCardActivity.finish()
        mProgressDialog.dismiss()
    }

}
